src/imitation/contrib/data_loader_wrappers.py

```diff
@@ -0,0 +1,211 @@
+
+    @abc.abstractmethod
+    def forwards_obs(
+        self,
+        obs: Any,
+    ) -> Any:
+        """Defines a batched transformation from raw to transformed observations.
+
+        Used transform observations returned by `gym.Wrapper.reset()`.
+        """
+
+    @abc.abstractmethod
+    def backwards_act(self, transformed_acts: np.ndarray):
+        """Defines a (pseudo-)inverse transformation from transformed act to raw act.
+
+        `backwards_act` should be a pure function. In other words, any instance of
+        BidirectionalTransform returns the same outputs given the same inputs.
+
+        Args:
+            transformed_acts: Batched actions transformed from raw actions by
+                `forwards()`.
+
+        Returns:
+            A raw action. Not required to be the same raw action as that passed into
+            `forwards()` because `forwards()` might not be one-to-one / injective.
+        """
+
+
+class BidirectionalDataLoaderWrapper(DataLoaderWrapper[Mapping, dict]):
+    """DataLoaderWrapper that corresponds to an instance of BidirectionalTransform.
+
+    Primarily used by BidirectionalGymWrapper.
+    """
+
+    def __init__(
+        self,
+        data_loader: types.DataLoaderInterface,
+        transform: BidirectionalTransform,
+    ):
+        self.data_loader = data_loader
+        self.transform = transform
+        super().__init__(data_loader)
+
+    def __iter__(self) -> dict:
+        for trans in self.data_loader:
+            result = self.transform.forwards(trans)
+            result["acts"] = self.transform.backwards_act(trans["acts"])
+            yield result
+
+
+class DefinesDataLoaderWrapper(abc.ABC, Generic[S, T]):
+    """Concrete instances of this abstract mixin define a `DatasetWrapper`.
+
+    Intended for use as an abstract mix-in in a subclass of `gym.Wrapper` so that the
+    `gym.Wrapper` can be processed by `wrap_data_loader_with_env_wrappers`.
+    """
+
+    @abc.abstractmethod
+    def apply_dataset_wrapper(
+        self,
+        data_loader: types.DataLoaderInterface,
+    ) -> DataLoaderWrapper[S, T]:
+        """Wraps a dataset using the dataset.
+
+        Args:
+            data_loader: A DataLoader to wrap.
+
+        Returns:
+            A wrapped data_loader.
+        """
+
+
+class BidirectionalGymWrapper(
+    gym.Wrapper,
+    DefinesDataLoaderWrapper[Mapping, dict],
+):
+    """Simple Gym Wrapper defined by an instance of BidirectionalTransform.
+
+    It trivially implements `DefinesDataLoaderWrapper.apply_data_loader_wrapper`
+    using the same `BidirectionalTransform`.
+    """
+
+    def __init__(self, env: gym.Env, transform: BidirectionalTransform):
+        self.transform = transform
+        super().__init__(env)
+
+    def reset(self):
+        obs = self.env.reset()
+        return self.forward_obs(obs)
+
+    def step(self, action: Any) -> Tuple[Any, float, dict]:
+        act = self.transform.backwards_act(action)
+        return self.env.step(act)
```

Comment:
This is missing a `forwards` step (for reward, observations, etc)?

Reply:
that's right, thanks